rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,364 @@
+
+## Required Approvers
+
+* Engineering: @russjones && @bernardjkim
+* Product: @klizhentas || @xinding33 
+* Security: @reedloden
+
+## What
+
+This RFD proposes a new mechanism for Teleport agents installed on Linux servers to automatically update to a version set by an operator via tctl.
+
+The following anti-goals are out-of-scope for this proposal, but will be addressed in future RFDs:
+- Analogous adjustments for Teleport agents installed on Kubernetes
+- Phased rollouts of new agent versions for agents connected to an existing cluster
+- Signing of agent artifacts via TUF
+- Teleport Cloud APIs for updating agents
+
+This RFD proposes a specific implementation of several sections in https://github.com/gravitational/teleport/pull/39217.
+
+Additionally, this RFD parallels the auto-update functionality for client tools proposed in https://github.com/gravitational/teleport/pull/39805.
+
+## Why
+
+The existing mechanism for automatic agent updates does not provide a hands-off experience for all Teleport users.
+
+1. The use of system package management leads to interactions with `apt upgrade`, `yum upgrade`, etc. that can result in unintentional upgrades or confusing command output.
+2. The use of system package management requires complex logic for each target distribution.
+3. The installation mechanism requires 4-5 commands, includes manually installing multiple packages, and varies depending on your version and edition of Teleport.
+4. The use of bash to implement the updater makes changes difficult and prone to error.
+5. The existing auto-updater has limited automated testing.
+6. The use of GPG keys in system package managers has key management implications that we would prefer to solve with TUF in the future.
+7. The desired agent version cannot be set via Teleport's operator-targeted CLI (tctl).
+8. The rollout plan for the new agent version is not fully-configurable using tctl.
+9. Agent installation logic is spread between the auto-updater script, install script, auto-discovery script, and documentation.
+10. Teleport contains logic that is specific to Teleport Cloud upgrade workflows.
+11. The existing auto-updater is not self-updating.
+12. It is difficult and undocumented to automate agent upgrades with custom automation (e.g., with JamF). 
+
+We must provide a seamless, hands-off experience for auto-updates that is easy to maintain.
+
+## Details
+
+We will ship a new auto-updater package written in Go that does not interface with the system package manager.
+It will be versioned separately from Teleport, and manage the installation of the correct Teleport agent version manually. 
+It will read the unauthenticated `/v1/webapi/ping` endpoint from the Teleport proxy, parse new fields on that endpoint, and install the specified agent version according to the specified upgrade plan.
+It will download the correct version of Teleport as a tarball, unpack it in `/var/lib/teleport`, and ensure it is symlinked from `/usr/local/bin`.
```

Comment:
I read "system software" as software installed and upgraded via an OS-provided tool like a package manager. I view these versions of teleport as installed by a user who invoked `teleport-updater enable` to create a local installation of Teleport.

I selected `/usr/local` over `/usr` to conform with:
> Software placed in / or /usr may be overwritten by system upgrades (though we recommend that distributions do not overwrite data in /etc under these circumstances). For this reason, local software must not be placed outside of /usr/local without good reason.

Seems better not to conflict at a file-level with the non-auto-updating `teleport` package?

Not strongly opinionated, but the OS won't be tracking these files via the package DB, so `/usr/local/bin` seems closer to the spec to me. Also, it's not unheard of for `/usr` to be read-only outside of system updates, and `/usr/local` to remain read-write.

https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s09.html

Reply:
> I read "system software" as software installed and upgraded via an OS-provided tool like a package manager

I agree. However this whole tool/endpoints/backing infra (on our end) _is_ a package manager system.

> I selected /usr/local over /usr to conform with:
> > Software placed in / or /usr may be overwritten by system upgrades

Right - which is exactly why we should place it there. This RFD is for part of a package management system, and one of the functions of this system is to provide Teleport upgrades.

> Seems better not to conflict at a file-level with the non-auto-updating teleport package, which is system software that has a binary in /usr/bin.

IMO in this case we should either:
* Advise Teleport admins to uninstall the normal Teleport package, or
* Forcefully uninstall Teleport installed by a package system manager, or
* Somehow adopt the current installation

> it's not unheard of for /usr to be read-only outside of system updates, and /usr/local to remain read-write

If customers complain about this then we could provide an override for installation directory via the config file. I haven't seen a ton of cases (not that they don't exist) of `/usr` being read only with `/usr/local` still writable.

Reply:
It might make sense to poll some customers and get them to weight in on this (and possibly some other updater implementation details).

Reply:
Hey folks, [this customer ticket](https://github.com/gravitational/teleport/issues/42130) just got added to Internal Tools goals for next quarter. Given that this shows at least one customer making the same argument that I am, I'm reopening this for discussion. I'd like to apply the same decision (one way or the other) to both this RFD and the customer's ticket.

Reply:
The customer is arguing that the teleport package should install the `teleport` binary to `/usr/bin` instead of `/usr/local/bin`, which I agree better conforms to the LSB spec.

Analogously, the upgrader installed by the upgrader package should live in `/usr/bin/teleport-updater`.

However, the teleport binaries that are installed by the user invoking `teleport-updater` commands are not managed by the system package manager, so it would not match the spec for them to land in `/usr/bin`.

Reply:
It doesn't matter whether a package comes from a package manager provided by the distro, or one that we write. To quote this customer's ticket:

> The FHS specifies that /usr/local is for manually installed software, not for packages managed by a package manager

We've established that Teleport's auto updates system is a package management system. `teleport-updater` is the local package manager. When the command is invoked, either manually or automatically (as is the primary intention of this work), it installs, upgrades, and removes `teleport` packages. This is not the user building Teleport themselves and `cp`ing the local build to a bin directory - this is our software managing, end to end, the lifecycle of our packages.

On top of all of that, the linked ticket shows that customers expect `teleport` to be in `/usr/bin`. It shouldn't matter to our customers _how_ they install Teleport on their OS. The behavior (including the location of files) should be the same, regardless of installation method (for a given OS). Customers are telling us that _they don't want `teleport` installed to this path_, so we shouldn't be installing it to this path regardless of installation method.

Reply:
> We've established that Teleport's auto updates system is a package management system.

I disagree that the FHS intends to suggest that any software that installs binaries counts as package manager. By this definition, `make` would be a packager manager, and `make install` should also install into `/usr/bin`.

> the linked ticket shows that customers expect teleport to be in /usr/bin

The linked ticket is referring to the Teleport system package, not a user-configured updater that installs Teleport from tarballs.

Reply:
> I disagree that the FHS intends to suggest that any software that installs binaries counts as package manager

I'm in agreement about this. FHS makes no mention of package managers at all. However by any reasonable definition of a package manager, `teleport-updater` is one, as a part of Teleport's auto updates package management system.

Here are several definitions of a "package manager" from several independent sources:
* Debian - https://www.debian.org/doc/manuals/aptitude/pr01s02.en.html
* Microsoft/Windows - https://learn.microsoft.com/en-us/windows/package-manager/
* National Instruments - https://www.ni.com/docs/en-US/bundle/package-manager/page/manual-overview.html
* Linode - https://www.linode.com/docs/guides/linux-package-management-overview/

By any of these definitions the auto updates system is a package management system, and the `teleport-updater` tool is a package manager.

> The linked ticket is referring to the Teleport system package, not a user-configured updater that installs Teleport from tarballs.

Of course you're correct here - it is in reference to the current Teleport package, and could not be in reference to a new system that has yet to be developed or deployed to customers. If you replace `RPM(\/| and )DEB` with `teleport-updater`, then it's clear that this issue will apply when the `teleport-updater` package manager is updating/installing `teleport` as well.